### ISSUE TITLE
Assert failure compiling constant if there are no other references to that type

### ISSUE THREAD

**Issue by cston** [dotnet/roslyn#13767]
Assert failure in `MetadataWriter.HeapOrReferenceIndexBase<T>.Add(T)` compiling: `csc /r:ValueTuple.dll /t:library c.cs`:
```
class C<T>
{
}
class C
{
    static void M()
    {
        const C<(int, int)> c = null;
    }
}
```

**Comment by jcouv** [comment]
This is not related to tuples. I can repro the assert with a referenced type which isn't `ValueTuple`.

```C#

        [Fact]
        [WorkItem(13767, "https://github.com/dotnet/roslyn/issues/13767")]
        public void GenericConstant()
        {
            var source = @"
class C<T>
{
}
class C
{
    static void M()
    {
        const C<System.ValueTupl<int, int>> c = null;
    }
}";

            var libSource = @"
namespace System
{
    public struct ValueTupl<T1, T2>
    {
        public ValueTupl(T1 item1, T2 item2) { }
    }
}
";
            var libComp = CreateCompilationWithMscorlib(libSource, assemblyName: "lib");
            libComp.VerifyDiagnostics();

            var comp = CreateCompilationWithMscorlib(source, assemblyName: "comp", references: new[] { libComp.EmitToImageReference() });
            comp.VerifyEmitDiagnostics();
        }
```

**Comment by jcouv** [comment]
Fixed by #13791
